cloudfoundry-incubator/cf-operator
# Add Bazel integration

## Description

Added Bazel build system integration for building the binary, building the Docker image in a deterministic way, publishing the image to the remote Docker registry and unit testing.

Left the old build system untouched.

The first time Bazel runs, it will download the dependencies and cache them in a very optimized way, making the next operations blazing fast. The invalidation of the cache happens on the change of the dependencies, causing a build on the reverse dependency tree. I.e. a change to a .go file will trigger the build of the targets that depend on the target that uses that file.

## Test plan

- The current build flow **SHOULD NOT** be affected.
- Check that all the Go targets are up-to-date with Gazelle: `bazel run //:gazelle`. No changes to the `BUILD.bazel` files should happen.
- Run the unit tests: `bazel test -- //... -//integration/... -//e2e/...`.
- Run the cf-operator on Minikube: `skaffold run --namespace scf`. Skaffold will trigger Bazel and then use the Helm charts to deploy. To validate the optimized cache, re-create the Minikube instance and perform another `skaffold run`. The cf-operator image should not be built again, only loaded into the Minikube Docker daemon (performed automatically by Skaffold).
- Build the image using `USE_BAZEL=1 bin/build-image`.
